### DISCUSSION TITLE
Big data

### DISCUSSION THREAD

**Discussion by deltamarnix** [pyphoenix/pyphoenix-project#49] [Ideas]
One of the bigger questions that we still have is how we are going to handle large datasets, also depending on the type of data (grid based / array based).

We should investigate how we can work with large data models where the data is chunked, preferably on disk. We see the benefits of using a package like Dask. Package that build on top of that are numpy, xarray, and uxarray. We are wondering if a package like uxarray also supports our more complex DISU model (fully unstructured 3d grid).

We have to consider how we are going to let users provide data to the flopy package. That can be via indexed arrays, but that can also be with a grid mask where NaNs are filled in when no data is provided. Depending on the sparsity of the data that is typically used, we can make a decision on the type of data we want users to put in via the API.

```py
import xarray as xa
import uxarray as uxa

class ModflowGwfriv(mfpackage.MFPackage):
    def __init__(self, stress_period_data: Union[xa.Grid, uxa.UnstructuredGrid, ...], ...):
        # If the river package requires information on all cells, it's likely best to accept a full grid
        pass
```

```py
import numpy as np

class  ModflowGwfriv(mfpackage.MFPackage):
    def __init__(self, stress_period_data: np.ArrayLike, ...):
        # This option would not give the user much information. The array is a 1-d array where the order of the data is important, like the current implementation. [cellid, stage, cond, rbot, aux, boundname]
        pass
```

```py
import numpy as np

class  ModflowGwfriv(mfpackage.MFPackage):
    def __init__(self, cellids: np.ArrayLike, stage: np.ArrayLike, cond: np.ArrayLike, rbot: np.ArrayLike, aux: np.ArrayLike, boundname: np.ArrayLike, ...):
        # As an alternative, the user could supply separate arrays and we combine it for the user and check that the lengths of these arrays are equal. It gives the user more information on what needs to go in, but the documentation has to be very clear on what is considered equal for the user.
        pass
```

```py
import pandas as pa

class  ModflowGwfriv(mfpackage.MFPackage):
    def __init__(self, stress_period_data: pa.DataFrame, ...):
        # Similar to a single np.ArrayLike, we could use pandas DataFrame so that the data is labelled and the order of the data is not relevant anymore.
        pass
```

Packages of interest:

- Xarray
- UXArray
- Zarr
- XUGrid
- Dask
- Numpy

XArray seems to support on disk formats for zarr and netcdf: https://docs.xarray.dev/en/stable/api.html#dataset-methods. Letting users open these data formats would only let us reference to it, perhaps write in it if needed. At the moment that the data needs to go to the MF6 input format, we can convert at that moment and not earlier. Making it more efficient.
If MF6 would support netcdf formats directly, there would be no conversion needed and it would save even more time.

**Comment by wpbonelli** [comment]
Just to note, MF6 distinguishes array and list data.

Arrays are always homogeneous and often (but not always) represent some variable living on features of the grid. I guess xarray would work only where this is true?

Lists can be homogeneous (elements all having the same record type) or heterogeneous (elements can be any of several record types). The former are amenable to numpy, pandas, providing columns separately, etc. The latter not so. Period data can be heterogeneous for some packages and so will not always be representable in a tabular format.

Also, for homogeneous arrays, maybe we could consider type hinting input context class attributes with e.g. `numpy.typing.NDArray[np.int_]` (and potentially providing a shape parameter in future, once numpy supports that). This would provide more info than simply `ArrayLike`, while the init method and other APIs could stay lenient and have an `ArrayLike` hint, and try to coerce the argument to an ndarray (or subclass / numpy-interoperable array class) of the proper type/shape. Kind of a best-of-both-worlds approach?

**Reply by deltamarnix** [reply]
Maybe we can work with `DataTree`? https://docs.xarray.dev/en/latest/user-guide/data-structures.html#datatree

**Reply by wpbonelli** [reply]
That is nice- especially the dimension inheritance. And it even has hierarchical addressing. It could represent the entire simulation context, at least all the space- or time-indexed variables, I guess?

Reminds me, we've talked about distinguishing "configuration" from "data" variables before, but I didn't have a clear idea where the line was. I am thinking now that if a variable is alignable with xarray it can be considered data? I'm not sure how useful a distinction this is, but maybe it has consequences, like which variables we can store in and/or present to user in a `DataTree`.

**Reply by wpbonelli** [reply]
some more notes on `DataTree`

- per the [docs](https://docs.xarray.dev/en/latest/user-guide/data-structures.html#coordinates), x, y, z, time etc would be dimensional coords, while kstp and kper could be non-dimensional coordinates
- we could use [custom accessors](https://docs.xarray.dev/en/stable/internals/extending-xarray.html#writing-custom-accessors) for e.g. plotting, layers, any information or functionality we want to attach to a dataset

**Comment by wpbonelli** [comment]
We've discussed/prototyped a modflow-specific [duck array](https://numpy.org/neps/nep-0022-ndarray-duck-typing-overview.html) type, supporting e.g. an optional multiplication factor, convenient layer access, and optimized storage for constant arrays. I wonder now if xarray can do at least the first two: layer as a labeled dimension and/or coordinate, factor stored as metadata and accessed/applied at write time.

**Reply by wpbonelli** [reply]
xarray [works with duck arrays](https://docs.xarray.dev/en/stable/user-guide/duckarrays.html) so any extra duck array types we provide would work with it

**Comment by wpbonelli** [comment]
Some questions we might be able to answer (with prototypes?) in the near term:

- [x] can xarray (via uxarray etc) represent DISV/DISU satisfactorily?
  - imod-python has disv https://deltares.github.io/imod-python/examples/mf6/circle.html#create-ugriddataarray
  - disu graph representation #58
- [x] what kind of improvements would xarray offer for large datasets?*
- [ ] why is flopy3 IO slow? is it at the write layer or further up the stack?*
- [x] can we use `DataTree` as an internal data store for the entire simulation context?
  - mentioned in https://github.com/modflowpy/pyphoenix-project/discussions/49#discussioncomment-11419724
  - prototyped in #57
- [ ] understand how flopy creates/maintains grids and dis/v/u packages

*These will require some benchmarking/profiling/performance testing, comparing flopy3 legacy IO with numpy/pandas/xarray IO

**Reply by deltamarnix** [reply]
I've researched some of the packages that seem to support UGRID. It's hard to find a library that supports 3D unstructured grids. I found the following so far:

- [UGridPy](https://github.com/Deltares/UGridPy), a Deltares product. Only supports 1D and 2D.
- [gridded](https://github.com/NOAA-ORR-ERD/gridded) is a package that only supports 2D grids (lat and lon). And it seems to store [everything in memory](https://noaa-orr-erd.github.io/gridded/gridded.pyugrid.html#module-gridded.pyugrid.ugrid).
- [XUGrid](https://github.com/Deltares/xugrid), a Deltares product by Huite. Only supports 2D, and can extend with adding an extra XArray dimension for `layer` in order to support DISV.
- [UXArray](https://github.com/UXARRAY/uxarray), currently being worked on by a masters student at NCAR. Supports 2D with XArray, so would likely also be able to add an extra dimension `layer` for DISV.

I haven't found any DISU compatible libraries. And I was thinking that it might be wise to keep the FloPy design as close to the MODFLOW6 input format as possible, like it is right now. So perhaps it's best as @mjreno suggested, to use numpy-like arrays so we can support dask where possible as input formats. Preferable as lazy arrays. And the moment that we need to write the MF6 files, we `compute()` the values and write them directly to the file format that is needed.

I will try to make a little prototype that shows how that works.

**Reply by deltamarnix** [reply]
I'm still playing in my head with the fact that DISU is actually just a graph, and should be treated that way. The graph itself might fit into memory, but the data on the nodes and edges might not.

There area some core packages in python that represent graphs: [python graphblas](https://pypi.org/project/python-graphblas/), [networkx](https://pypi.org/project/networkx/), [scipy](https://pypi.org/project/scipy/). And on top of that there is a package that can translate graphs from one to another, or call algorithms from one on another type: [metagraph](https://pypi.org/project/metagraph/) (metagraph was last released in 2021, so not as interesting anymore).

So far we want to stay as close as possible with the packages to the input of modflow, but we might want to use these packages in generator helper functions networkx -> gwfdisu. Or we might accept a networkx-graph as input if it is set up correctly and convert it to our own representation internally.

For the output we can do something similar. By default we return a single list of data points based on the cell index, but if there are use cases for traversal, we can add a translation layer on top to return a networkx or graphblas object with the data linked to the correct nodes and edges.

**Reply by deltamarnix** [reply]
I'm diving a little deeper into what XArray supports for representing a graph (DISU) in its data format. [XArray](https://docs.xarray.dev/en/latest/user-guide/duckarrays.html) supports [numpy sparse arrays](https://sparse.pydata.org/en/stable/).

NetworkX has the option to transform its representation to numpy arrays [to_numpy_array](https://networkx.org/documentation/stable/reference/generated/networkx.convert_matrix.to_numpy_array.html). This information can be directly stored in an XArray object. And the other way around, a network representation in XArray can be transformed back to a NetworkX object with [from_numpy_array](https://networkx.org/documentation/stable/reference/generated/networkx.convert_matrix.from_numpy_array.html).

**Reply by deltamarnix** [reply]
I made a little prototype in #58 to show how we could store networkx graph via a sparse scipy matrix into an XArray Dataset, and get the structure back into networkx too.

**Comment by wpbonelli** [comment]
Another question: how to lazy load large arrays with cattrs? Or, how to use another more suitable library and use cattrs only for configuration input?

**Comment by wpbonelli** [comment]
a question: should we send all array data through xarray/dask and do lazy/chunked operations on it, or operate on data below a certain size in memory?

**Comment by deltamarnix** [comment]
We were discussing how XArray and CAttrs would work together like in this PR: #62 .

cattrs will always return a dictionary of some sort, so that it can be passed into packages like pyyaml or tomlkit. cattrs provides the conversion for datetimes and encoding. But cattrs is not made for file-io, because when you would want to convert everything into one big string, that would all be passed into memory in a single instance.

It would be better to use functionality like jinja for file writing and just pass it the actual python instance of simulation / model / package. We can write filter functions that can handle the dask chunking while writing large datasets from a netcdf file / lazy dask array. This is something we still need to prototype.

**Reply by wpbonelli** [reply]
separate discussion in #71

**Comment by deltamarnix** [comment]
I have worked on profiling the different methods to write ascii input files to disk to find what is fastest and doesn't take up too much memory.

I came up with the following functions:

- `xarray-extras`: A small library with a C++ implementation for fast writing of xarray to csv. And csv is what we want, in a way, but with spaces instead of commas.
- `xarray.to_pandas().to_csv()`: The built-in function to write to csv files.
- `np.save_txt()`: A simple function from numpy. We call this function block for block via xarray.
- `jinja` calling `np.array2string()`: A similar function to `save_txt`, but not to a file handle.

Here are the results:

| test | args | peak memory (MB) | time (s) |
| --- | --- | --- | --- |
| create_and_write_extras | {'max_size': 1000000, 'chunks': 10000} | 18.941005 MB | 0.6280782222747803 |
| create_and_write_extras | {'max_size': 10000000, 'chunks': 10000} | 162.426803 MB | 4.197936773300171 |
| create_and_write_pandas | {'max_size': 1000000, 'chunks': 10000} | 17.273536 MB | 0.6013274192810059 |
| create_and_write_pandas | {'max_size': 10000000, 'chunks': 10000} | 164.558039 MB | 8.126498222351074 |
| create_and_write_np_savetxt | {'max_size': 1000000, 'chunks': 10000} | 8.641115 MB | 3.409224033355713 |
| create_and_write_np_savetxt | {'max_size': 10000000, 'chunks': 10000} | 83.415583 MB | 43.03009843826294 |
| create_and_write_jinja | {'max_size': 1000000, 'chunks': 10000} | 8.649213 MB | 0.44498276710510254 |
| create_and_write_jinja | {'max_size': 10000000, 'chunks': 10000} | 83.400212 MB | 21.505234479904175 |

We can see that xarray-extras takes around twice as much space for larger datasets at peak. But it is also 10x faster.

I am also running tests for 1 billion points. More results later.

**Reply by deltamarnix** [reply]
I did a 1 billion point test with `create_and_write_extras`, because it seems to be the most viable option:

| test | args | memory (MB) | time (s) |
| --- | --- | --- | --- |
| create_and_write_extras | {'max_size': 1000000000, 'chunks': 10000000} | 8000.121117 MB | 175.37805652618408 |

So yes, there still seems to be quite some data increase when we do a big chunk of data. But we can also state that writing 1B points as ascii is not so realistic, and we should be looking into binary dumping instead.

**Reply by deltamarnix** [reply]
More test results

| test | args | memory (MB) | time (s) |
| --- | --- | --- | --- |
| create_and_write_extras | {'max_size': 1000, 'chunks': 100} | 0.434587 MB | 0.23753690719604492 |
| create_and_write_extras | {'max_size': 1000000, 'chunks': 1000} | 23.305159 MB | 3.632298469543457 |
| create_and_write_extras | {'max_size': 10000000, 'chunks': 10000} | 159.64868 MB | 5.064882516860962 |
| create_and_write_extras | {'max_size': 1000000000, 'chunks': 10000000} | 8000.076105 MB | 144.34528183937073 |
| create_and_write_pandas | {'max_size': 1000, 'chunks': 100} | 0.32561 MB | 0.016986370086669922 |
| create_and_write_pandas | {'max_size': 1000000, 'chunks': 1000} | 20.328459 MB | 1.1458525657653809 |
| create_and_write_pandas | {'max_size': 10000000, 'chunks': 10000} | 164.252239 MB | 8.57870602607727 |
| create_and_write_pandas | {'max_size': 1000000000, 'chunks': 10000000} | 18170.165632 MB | 980.215124130249 |
| create_and_write_np_savetxt | {'max_size': 1000, 'chunks': 100} | 0.192303 MB | 0.10468173027038574 |
| create_and_write_np_savetxt | {'max_size': 1000000, 'chunks': 1000} | 11.274385 MB | 11.107773542404175 |
| create_and_write_np_savetxt | {'max_size': 10000000, 'chunks': 10000} | 83.435863 MB | 52.098299980163574 |
| create_and_write_np_savetxt | {'max_size': 1000000000, 'chunks': 10000000} | did not finish | 2850.3509786129 |
| create_and_write_jinja | {'max_size': 1000, 'chunks': 100} | 0.214198 MB | 0.024641036987304688 |
| create_and_write_jinja | {'max_size': 1000000, 'chunks': 1000} | 11.705371 MB | 7.205366611480713 |
| create_and_write_jinja | {'max_size': 10000000, 'chunks': 10000} | 83.926305 MB | 24.332927227020264 |
| create_and_write_jinja | {'max_size': 1000000000, 'chunks': 10000000} | 483.047364 MB | 6.90538477897644 (data gets written in ellipsis notation, so wrong) |

**Reply by deltamarnix** [reply]
My advice would be to go for `xarray-extras` for ascii writing. The package is a fairly simple C++ implementation. It's faster than all other options I have explored, and the memory usage is doable.

One limitation would be that it writes all values onto one line. We have no control over that.

**Reply by deltamarnix** [reply]
I also found out that imod-python is using the `pandas.to_csv()` function as already tested here for writing a disv schematization.

https://github.com/Deltares/imod-python/blob/eb7998a36b1bec47e589b6e4cb4ccf2b6cb2cb80/imod/mf6/disv.py#L110

https://github.com/Deltares/imod-python/blob/eb7998a36b1bec47e589b6e4cb4ccf2b6cb2cb80/imod/mf6/disv.py#L141-L143

**Reply by deltamarnix** [reply]
I also did some tests with binary writing

| test | args | memory (MB) | time (s) |
| --- | --- | --- | --- |
| create_and_write_flatten | {'max_size': 1000000000, 'chunks': 10000000} | Out of memory exception | Out of memory exception |
| create_and_write_flatten_blocks | {'max_size': 1000000000, 'chunks': 10000000} | 483.090754 MB | 30.942705392837524 |

**Reply by deltamarnix** [reply]
@mjreno I have a very simple test for netcdf with xarray. It's a simple `xarray.to_netcdf()`.

| test | args | time (s) |
| --- | --- | --- |
| create_and_write_netcdf | {'max_size': 1000000000, 'chunks': 10000000} | 107.73747897148132 |

**Reply by deltamarnix** [reply]
I did a 2D test with 1_000_000_000 values reshaped: 

| test | args | time (s) |
| --- | --- | --- |
| create_and_write_extras_2d | {'shape': (25000, 40000)} | 125.32499647140503 |

So that's similar to 1D data.

**Comment by jdhughes-dev** [comment]
Maybe we could set some defaults that writes data to binary files when the size exceeds a certain size and provide an override option for users that want to fill their hard drive with ascii data.

**Comment by deltamarnix** [comment]
We can conclude that it's possible to make larger models by keeping data on disk as long as possible, meaning we want to delay the calculations until the moment of writing the MODFLOW input files. We can achieve this via dask arrays that can be provided to components, instead of numpy in-memory arrays.

We have done some tests on different writing methods and we see that there are external libraries (`xarray-extras`) that are efficient, but still take a large chunk of memory. The built-in xarray methods for writing can be more memory efficient than external libraries like `xarray-extras`, but take longer to write MODFLOW input files. Both 1D and 2D data has a similar performance.

We see that large chunks of data can be written more efficient with binary writers. We have to make a decision on where we can draw that boundary switching between binary and ascii files.
